File: backend_flask/routes/api_keys.py
import hashlib
import secrets

from flask import Blueprint, g, jsonify, request

from ..auth_utils import authenticate_token
from ..extensions import db
from ..models import ApiKey, utcnow

api_keys_bp = Blueprint("api_keys", __name__, url_prefix="/api/api-keys")


def hash_key(key):
    return hashlib.sha256(key.encode("utf-8")).hexdigest()


@api_keys_bp.get("/")
@authenticate_token
def list_api_keys():
    keys = (
        ApiKey.query
        .filter_by(user_id=g.user.get("id"))
        .order_by(ApiKey.created_at.desc())
        .all()
    )
    return jsonify({"data": [key.to_public_dict() for key in keys]})


@api_keys_bp.post("/")
@authenticate_token
def create_api_key():
    data = request.get_json(silent=True) or {}
    name = str(data.get("name") or "HireFlow API Key").strip() or "HireFlow API Key"
    raw_key = f"hf_{secrets.token_urlsafe(24)}"

    api_key = ApiKey(
        user_id=g.user.get("id"),
        name=name,
        key_hash=hash_key(raw_key),
        key_prefix=f"{raw_key[:8]}...",
    )
    db.session.add(api_key)
    db.session.commit()

    payload = api_key.to_public_dict()
    payload["key"] = raw_key
    return jsonify({
        "message": "API key created. Copy it now; it will not be shown again.",
        "data": payload,
    }), 201


@api_keys_bp.delete("/<key_id>")
@authenticate_token
def revoke_api_key(key_id):
    try:
        numeric_id = int(key_id)
    except (TypeError, ValueError):
        return jsonify({"error": "Invalid API key ID"}), 400

    api_key = (
        ApiKey.query
        .filter_by(id=numeric_id, user_id=g.user.get("id"), revoked_at=None)
        .first()
    )
    if not api_key:
        return jsonify({"error": "API key not found"}), 404

    api_key.revoked_at = utcnow()
    db.session.commit()
    return jsonify({"message": "API key revoked successfully"})

File: backend/src/routes/apiKeys.ts
import { Router } from 'express';
import { ApiKeyModel } from '../models/ApiKey.ts';
import { authenticateToken } from '../middleware/auth.ts';

const router = Router();
let apiKeyModel: ApiKeyModel;

export function setApiKeyModel(db: any) {
  apiKeyModel = ApiKeyModel.create(db);
}

router.get('/', authenticateToken, async (req, res) => {
  try {
    const keys = await apiKeyModel.listByUser(req.user!.id);
    res.json({ data: keys });
  } catch (error) {
    console.error('List API keys error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/', authenticateToken, async (req, res) => {
  try {
    const name = typeof req.body?.name === 'string' ? req.body.name : 'HireFlow API Key';
    const key = await apiKeyModel.create(req.user!.id, name);
    res.status(201).json({
      message: 'API key created. Copy it now; it will not be shown again.',
      data: key,
    });
  } catch (error) {
    console.error('Create API key error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.delete('/:id', authenticateToken, async (req, res) => {
  try {
    const id = Number(req.params.id);
    if (!Number.isFinite(id)) {
      return res.status(400).json({ error: 'Invalid API key ID' });
    }

    const revoked = await apiKeyModel.revoke(req.user!.id, id);
    if (!revoked) {
      return res.status(404).json({ error: 'API key not found' });
    }

    res.json({ message: 'API key revoked successfully' });
  } catch (error) {
    console.error('Revoke API key error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;
